### ISSUE TITLE
Each ios_config command opens a new SSH session without closing it when it's done

### ISSUE THREAD

**Issue by jean-christophe-manciot** [ansible/ansible-modules-core#5181]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

ios_config
##### ANSIBLE VERSION

```
ansible 2.2.0.0 RC1
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

inventory   = ./hosts
gathering = explicit
roles_path = /home/actionmystique/Program-Files/Ubuntu/Ansible/git-Ansible/roles
private_role_vars = yes
log_path = /var/log/ansible.log
fact_caching = redis
fact_caching_timeout = 86400
retry_files_enabled = False
##### OS / ENVIRONMENT
- host: Ubuntu 16.04 4.4.0
- target: **CSR-1000v 16.2.2**
##### SUMMARY

ios_config executes multiple commands correctly, opening a new SSH session for each command **without closing the previous one** until the target runs out of sessions (16 IN and 16 OUT) and refuses any new session, hence triggering an "Unable to connect to port 22" on ansible side.
This behavior happens only with CSR-1000v and not with other Cisco devices (so far).
##### STEPS TO REPRODUCE

**Inventory ./hosts**:

```
[spines_sf]
XEv_Spine_31 ansible_host=172.21.100.231 os_family=ios_xev
```

Structure passed as "**provider**": connections.ssh
Defined in group_vars/ios_xev/connections.yml and a symbolic link in roles/ios_snmp/defaults points to ios_xev

```
connections
...
        ssh:
          host: "{{ inventory_hostname }}"
          port: 22
          username: xxxxx
          password: xxxxxxxx
          authorize: yes
          auth_pass: xxxxxxxx
          ssh_keyfile: "~/.ssh/id_rsa"
          timeout: 10
```

**snmp** structure defined in group_vars/ios_xev/snmp.yml and a symbolic link in roles/ios_snmp/defaults points to ios_xev:

```
snmp:
        admin_state: up
...
        new_version: 2c
...
        notifications_type: traps
        notifications: list
        notifications_list:
          traps:
                - bfd
                - bgp
                - cpu
                - eigrp
                - event-manager
                - firewall serverstatus
                - ike policy add
...
```

**Role**: ios_snmp:

```
- include_vars: "../defaults/{{ os_family }}/connections.yml"
- include_vars: "../defaults/{{ os_family }}/snmp.yml"
...
- name: Configuring the list of notifications in SNMPv{{ snmp.new_version }}
  ios_config:
        provider: "{{ connections.ssh }}"
        lines:
          - "snmp-server enable traps {{ item }}"
  with_items:  "{{ snmp.notifications_list.traps }}"
  when: (snmp.notifications == 'list')
  register: result
```

**Playbook**:

```
- name: Configuring SNMP on IOS/IOSv/IOSv-L2/IOS-XE/IOS-XEv
  hosts:
# -------------Buggy for ios: https://github.com/ansible/ansible/issues/16546----------------------
    - ios
    - iosv
    - iosv_l2
    - ios_xev
  roles:
    - ios_snmp
```
##### EXPECTED RESULTS

The SSH sessions should be closed after each ios_config or preferably at the end of the workflow without opening a new session for each command.
##### ACTUAL RESULTS

```
...
TASK [ios_snmp : Configuring location, contact & chassis ID information] *******
changed: [XEv_Spine_31] => {"changed": true, "updates": ["snmp-server location California", "snmp-server contact manciot.jeanchristophe@gmail.com", "snmp-server chassis-id 1", "snmp-server ifindex persist"], "warnings": []}
changed: [XEv_Spine_32] => {"changed": true, "updates": ["snmp-server location California", "snmp-server contact manciot.jeanchristophe@gmail.com", "snmp-server chassis-id 1", "snmp-server ifindex persist"], "warnings": []}

TASK [ios_snmp : Configuring the ACL for authorized SNMP NMS] ******************
changed: [XEv_Spine_32] => {"changed": true, "updates": ["ip access-list extended authorized-snmp-nms", "permit ip host 172.21.100.1 host 172.21.100.232 log"], "warnings": []}
changed: [XEv_Spine_31] => {"changed": true, "updates": ["ip access-list extended authorized-snmp-nms", "permit ip host 172.21.100.1 host 172.21.100.231 log"], "warnings": []}

TASK [ios_snmp : Disabling all notifications in SNMPv3] ************************
changed: [XEv_Spine_31] => {"changed": true, "updates": ["no snmp-server host 172.21.100.1 traps version 3 priv admin"], "warnings": []}
changed: [XEv_Spine_32] => {"changed": true, "updates": ["no snmp-server host 172.21.100.1 traps version 3 priv admin"], "warnings": []}

TASK [ios_snmp : Disabling the user name, its authentication/encryption parameters & the allowed NMS in SNMPv3] ***
changed: [XEv_Spine_32] => {"changed": true, "updates": ["no snmp-server user admin admin v3  auth sha ******** priv aes 192 ********"], "warnings": []}
changed: [XEv_Spine_31] => {"changed": true, "updates": ["no snmp-server user admin admin v3  auth sha ******** priv aes 192 ********"], "warnings": []}

TASK [ios_snmp : Disabling the group name, its security level & the allowed NMS in IPv4/SNMPv3] ***
changed: [XEv_Spine_31] => {"changed": true, "updates": ["no snmp-server group admin v3 priv access authorized-snmp-nms"], "warnings": []}
changed: [XEv_Spine_32] => {"changed": true, "updates": ["no snmp-server group admin v3 priv access authorized-snmp-nms"], "warnings": []}

TASK [ios_snmp : Configuring RO/RW community string in IPv4/SNMPv2c] ***********
changed: [XEv_Spine_31] => {"changed": true, "updates": ["snmp-server community Iosv2016$ ro authorized-snmp-nms"], "warnings": []}
changed: [XEv_Spine_32] => {"changed": true, "updates": ["snmp-server community Iosv2016$ ro authorized-snmp-nms"], "warnings": []}

TASK [ios_snmp : Configuring which NMS is allowed to receive the traps in SNMPv2c] ***
changed: [XEv_Spine_31] => {"changed": true, "updates": ["snmp-server host 172.21.100.1 traps version 2c Iosv2016$ udp-port 162"], "warnings": []}
changed: [XEv_Spine_32] => {"changed": true, "updates": ["snmp-server host 172.21.100.1 traps version 2c Iosv2016$ udp-port 162"], "warnings": []}

TASK [ios_snmp : Configuring the list of notifications in SNMPv2c] *************
changed: [XEv_Spine_32] => (item=bfd) => {"changed": true, "item": "bfd", "updates": ["snmp-server enable traps bfd"], "warnings": []}
changed: [XEv_Spine_31] => (item=bfd) => {"changed": true, "item": "bfd", "updates": ["snmp-server enable traps bfd"], "warnings": []}
changed: [XEv_Spine_31] => (item=bgp) => {"changed": true, "item": "bgp", "updates": ["snmp-server enable traps bgp"], "warnings": []}
changed: [XEv_Spine_32] => (item=bgp) => {"changed": true, "item": "bgp", "updates": ["snmp-server enable traps bgp"], "warnings": []}
changed: [XEv_Spine_32] => (item=cpu) => {"changed": true, "item": "cpu", "updates": ["snmp-server enable traps cpu"], "warnings": []}
changed: [XEv_Spine_31] => (item=cpu) => {"changed": true, "item": "cpu", "updates": ["snmp-server enable traps cpu"], "warnings": []}
changed: [XEv_Spine_32] => (item=eigrp) => {"changed": true, "item": "eigrp", "updates": ["snmp-server enable traps eigrp"], "warnings": []}
changed: [XEv_Spine_31] => (item=eigrp) => {"changed": true, "item": "eigrp", "updates": ["snmp-server enable traps eigrp"], "warnings": []}
changed: [XEv_Spine_32] => (item=event-manager) => {"changed": true, "item": "event-manager", "updates": ["snmp-server enable traps event-manager"], "warnings": []}
changed: [XEv_Spine_31] => (item=event-manager) => {"changed": true, "item": "event-manager", "updates": ["snmp-server enable traps event-manager"], "warnings": []}
changed: [XEv_Spine_31] => (item=firewall serverstatus) => {"changed": true, "item": "firewall serverstatus", "updates": ["snmp-server enable traps firewall serverstatus"], "warnings": []}
changed: [XEv_Spine_32] => (item=firewall serverstatus) => {"changed": true, "item": "firewall serverstatus", "updates": ["snmp-server enable traps firewall serverstatus"], "warnings": []}
changed: [XEv_Spine_31] => (item=ike policy add) => {"changed": true, "item": "ike policy add", "updates": ["snmp-server enable traps ike policy add"], "warnings": []}
changed: [XEv_Spine_32] => (item=ike policy add) => {"changed": true, "item": "ike policy add", "updates": ["snmp-server enable traps ike policy add"], "warnings": []}
changed: [XEv_Spine_31] => (item=ike policy delete) => {"changed": true, "item": "ike policy delete", "updates": ["snmp-server enable traps ike policy delete"], "warnings": []}
changed: [XEv_Spine_32] => (item=ike policy delete) => {"changed": true, "item": "ike policy delete", "updates": ["snmp-server enable traps ike policy delete"], "warnings": []}
changed: [XEv_Spine_31] => (item=ike tunnel start) => {"changed": true, "item": "ike tunnel start", "updates": ["snmp-server enable traps ike tunnel start"], "warnings": []}
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: paramiko.ssh_exception.NoValidConnectionsError: [Errno None] Unable to connect to port 22 on 172.21.100.231
failed: [XEv_Spine_31] (item=ike tunnel stop) => {"changed": false, "failed": true, "item": "ike tunnel stop", "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_BAaxb0/ansible_module_ios_config.py\", line 362, in <module>\n    main()\n  File \"/tmp/ansible_BAaxb0/ansible_module_ios_config.py\", line 352, in main\n    run(module, result)\n  File \"/tmp/ansible_BAaxb0/ansible_module_ios_config.py\", line 263, in run\n    config = get_config(module, result)\n  File \"/tmp/ansible_BAaxb0/ansible_module_ios_config.py\", line 232, in get_config\n    contents = module.config.get_config(include_defaults=defaults)\n  File \"/tmp/ansible_BAaxb0/ansible_modlib.zip/ansible/module_utils/network.py\", line 125, in config\n  File \"/tmp/ansible_BAaxb0/ansible_modlib.zip/ansible/module_utils/network.py\", line 147, in connect\n  File \"/tmp/ansible_BAaxb0/ansible_modlib.zip/ansible/module_utils/ios.py\", line 180, in connect\n  File \"/tmp/ansible_BAaxb0/ansible_modlib.zip/ansible/module_utils/shell.py\", line 228, in connect\n  File \"/tmp/ansible_BAaxb0/ansible_modlib.zip/ansible/module_utils/shell.py\", line 100, in open\n  File \"/usr/local/lib/python2.7/dist-packages/paramiko/client.py\", line 324, in connect\n    raise NoValidConnectionsError(errors)\nparamiko.ssh_exception.NoValidConnectionsError: [Errno None] Unable to connect to port 22 on 172.21.100.231\n", "module_stdout": "", "msg": "MODULE FAILURE"}
```

On Cisco target, the SSH sessions are all still open:

```
XEv_Spine_31#sh ssh
Connection Version Mode Encryption  Hmac         State                 Username
0          2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
0          2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
1          2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
1          2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
2          2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
2          2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
3          2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
3          2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
4          2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
4          2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
5          2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
5          2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
6          2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
6          2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
7          2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
7          2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
8          2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
8          2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
9          2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
9          2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
10         2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
10         2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
11         2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
11         2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
12         2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
12         2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
13         2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
13         2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
14         2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
14         2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
15         2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
15         2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
XEv_Spine_31#
```

On **IOSv, only 2 SSH sessions remain open** during the whole identical workflow:

```
Level_3#sh ssh
Connection Version Mode Encryption  Hmac         State                 Username
0          2.0     IN   aes128-ctr  hmac-sha1-96 Session started       admin
0          2.0     OUT  aes128-ctr  hmac-sha1-96 Session started       admin
%No SSHv1 server connections running.
Level_3#
```

**Comment by ansibot** [comment]
@privateip, @gundalow, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by gundalow** [comment]
Thank you for the bug report and detailed analysis. 

We will review this after AnsibleFest, and get back to you.

**Comment by privateip** [comment]
@jean-christophe-manciot i just committed a change that should address this problem.  The ios_config module was definitely not explicitly disconnecting from the remote device.  Unfortunately I dont have copy of 16.2 and was unable to replicate the problem on 16.3.  Therefore I'm leaving this issue open until you have a chance to test again and confirm the fix works.

**Comment by privateip** [comment]
needs_info

**Comment by jean-christophe-manciot** [comment]
This issue has been solved & merged into ansible-modules-core but the ref to the submodules has not been updated yet.

**Comment by privateip** [comment]
fixed and verified
